### PR TITLE
Make ConsignmentStatus' statusValue optional but only for 'Upload'

### PR DESCRIPTION
**Pull request by techncl**
Instead of creating a new endpoint I adapted the updateConsignmentStatus to take 'statusValue' as an optional parameter. Although it's optional, it will only be optional for 'Upload' for now (there is logic to check this). This way, when the files are uploaded, the frontend can call updateConsignmentStatus and then whether the upload is 'Complete' or 'CompletedWithIssues' can be determined by the API (via the fileStatuses) rather than the frontend

### CODE REVIEW

**Review by TomJKing — COMMENTED** [review]
Having a bit of the think about this, I'm wondering if its better to continue to pass in the "status value", but ignore it for those consignment status updates we want to generate from the file statuses. I've put an outline of this approach on a branch: https://github.com/nationalarchives/tdr-consignment-api/tree/consignment-statuses-example

In my head this will give the following advantages:

- not break any existing calls to the update statuses endpoint
- allow the migration of other consignment statuses over to the new approach with minimal code changes
- can then retire the old endpoint once everything is migrated

Let me know what you think.

File: schema.graphql [hunk 1]
```diff
@@ -157,7 +157,7 @@ type ConsignmentStatus {
 input ConsignmentStatusInput {
   consignmentId: UUID!
   statusType: String!
-  statusValue: String!
+  statusValue: String
```
[the remark this answers is not in the document]
Comment (techncl) [reply schema.graphql:160]: I don't think so because I left it out in the RouteSpec tests [here](https://github.com/nationalarchives/tdr-consignment-api/pull/449/files#diff-9c6f7642b0572956c29857db1de71fe336af0c726f5af176d97afb7bb93fe2db). I think if you pass a value, Sangria  will wrap it in a `Some` and if you don't, it will just make it a `None`

**Comment by techncl** [comment]
> Having a bit of the think about this, I'm wondering if its better to continue to pass in the "status value", but ignore it for those consignment status updates we want to generate from the file statuses. I've put an outline of this approach on a branch: https://github.com/nationalarchives/tdr-consignment-api/tree/consignment-statuses-example
> 
> In my head this will give the following advantages:
> 
> * not break any existing calls to the update statuses endpoint
> * allow the migration of other consignment statuses over to the new approach with minimal code changes
> * can then retire the old endpoint once everything is migrated
> 
> Let me know what you think.

I like that your solution doesn't have to worry about checking for a missing `statusValue` for non-Upload statusTypes, I was thinking "what about when we need to set the Upload `statusValue` to `InProgress`?" but then realised that we are calling the repo method directly to set that so as long as we don't switch to calling the service, we should be fine.

Wouldn't making `statusValue` optional make migration easier since you don't need to pass it in?

File: schema.graphql (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomJKing) [reply schema.graphql:160]: Ah OK, for once something useful. So yeah, makes sense to make it optional.

**Comment by TomJKing** [comment]
> ... I was thinking "what about when we need to set the Upload statusValue to InProgress?" but then realised that we are calling the repo method directly to set that so as long as we don't switch to calling the service, we should be fine.

You're right about setting the "Inprogress" status, its a tricky one. I think we can leave it is for now, and maybe revisit it if needs be.

> Wouldn't making statusValue optional make migration easier since you don't need to pass it in?

Yes, I was thinking that setting it to optional would break any calls to the endpoint, but as you pointed out this should be OK.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusService.scala [hunk 2]
```diff
@@ -67,18 +70,34 @@ class ConsignmentStatusService(consignmentStatusRepository: ConsignmentStatusRep
...
+        updateConsignmentStatusInput.consignmentId,
+        updateConsignmentStatusInput.statusType,
+        updateConsignmentStatusInput.statusValue.get, // if this value is None, then it should through an error as this should not be possible after the validation
+        Timestamp.from(timeSource.now)
+      )
+    }
+  }
+
+  private def updateBasedOnFileStatus(consignmentId: UUID, statusType: String): Future[Int] = {
+    val noFileStatusesError = s"Error: There are no $statusType statuses for any of the files from consignment $consignmentId"
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusService.scala:89]: Maybe re-word error message slightly to: `... statuses for any files from consignment ...`

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusService.scala [hunk 3]
```diff
@@ -67,18 +70,34 @@ class ConsignmentStatusService(consignmentStatusRepository: ConsignmentStatusRep
...
+    val statusType: String = updateConsignmentStatusInput.statusType
+
+    if (statusesToUpdateBasedOnFile.contains(statusType)) {
+      updateBasedOnFileStatus(updateConsignmentStatusInput.consignmentId, statusType)
+    } else {
+      validateStatusTypeAndValue(updateConsignmentStatusInput)
+      consignmentStatusRepository.updateConsignmentStatus(
+        updateConsignmentStatusInput.consignmentId,
+        updateConsignmentStatusInput.statusType,
+        updateConsignmentStatusInput.statusValue.get, // if this value is None, then it should through an error as this should not be possible after the validation
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusService.scala:82]: I would remove this comment.

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentStatusRouteSpec.scala [hunk 4]
```diff
@@ -254,4 +254,91 @@ class ConsignmentStatusRouteSpec extends TestContainerUtils with Matchers with T
...
+      utils.createConsignment(consignmentId, userId)
+
+      val expectedResponse = expectedUpdateConsignmentStatusMutationResponse("data_missing_statusvalue_not_allowed")
+      val response = runUpdateConsignmentStatusTestMutation("mutation_missing_statusvalue_not_allowed", token)
+
+      response.errors.head.message should equal(expectedResponse.errors.head.message)
+      response.errors.head.extensions should equal(expectedResponse.errors.head.extensions)
+  }
+
+  "updateConsignmentStatus" should "return an error if statusType is 'Upload', statusValue wasn't passed in but no files have 'Upload' statuses" in withContainers {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentStatusRouteSpec.scala:273]: Maybe reword test title slightly to:` ... if statusType is 'Upload' and no statusValue is passed where consignment has no files with 'Upload' statuses`

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentStatusRouteSpec.scala [hunk 5]
```diff
@@ -254,4 +254,91 @@ class ConsignmentStatusRouteSpec extends TestContainerUtils with Matchers with T
...
+      utils.createConsignmentStatus(consignmentId, statusType, statusValue)
+
+      val expectedResponse = expectedUpdateConsignmentStatusMutationResponse("data_no_files_uploaded")
+      val response = runUpdateConsignmentStatusTestMutation("mutation_no_files_uploaded", token)
+
+      response.errors.head.message should equal(expectedResponse.errors.head.message)
+      response.errors.head.extensions should equal(expectedResponse.errors.head.extensions)
+  }
+
+  "updateConsignmentStatus" should "update the consignment status if statusType is 'Upload', statusValue wasn't passed in " +
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentStatusRouteSpec.scala:294]: Maybe reword test title slightly to: `... if statusType is 'Upload' and no statusValue was passed in where all consignment files have 'Success' Upload statuses`

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentStatusRouteSpec.scala [hunk 6]
```diff
@@ -254,4 +254,91 @@ class ConsignmentStatusRouteSpec extends TestContainerUtils with Matchers with T
...
+      utils.createFileStatusValues(UUID.randomUUID(), fileId1, statusType, "Success")
+      utils.createFileStatusValues(UUID.randomUUID(), fileId2, statusType, "Success")
+
+      val expectedResponse = expectedUpdateConsignmentStatusMutationResponse("data_all")
+      val response = runUpdateConsignmentStatusTestMutation("mutation_files_all_uploaded_successfully", token)
+
+      response.data.get.updateConsignmentStatus should equal(expectedResponse.data.get.updateConsignmentStatus)
+  }
+
+  "updateConsignmentStatus" should "update the consignment status if statusType is 'Upload', statusValue wasn't passed in " +
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentStatusRouteSpec.scala:320]: Maybe reword test title slightly to: `... if statusType is 'Upload' and no statusValue was passed in where some consignment files have 'Success' Upload statuses`

File: src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala [hunk 7]
```diff
@@ -350,13 +380,261 @@ class ConsignmentStatusServiceSpec extends AnyFlatSpec with MockitoSugar with Re
...
       consignmentService.updateConsignmentStatus(updateConsignmentStatusInput).futureValue
     }
 
-    thrownException.getMessage should equal(s"Invalid ConsignmentStatus input: either '$expectedStatusType' or '$expectedStatusValue'")
+    thrownException.getMessage should equal(
+        s"Invalid ConsignmentStatus input: either '$expectedStatusType' or '$expectedStatusValue'"
+    )
+  }
+
+  "updateConsignmentStatus" should s"call fileStatus repo in order to generate a statusValue when only a consignmentId and 'Upload' statusType have been passed in" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala:394]: Maybe re-word test title slightly to: `... repo to generate a statusValue when a consignmentId and 'Upload' ...`

File: src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala [hunk 8]
```diff
@@ -350,13 +380,261 @@ class ConsignmentStatusServiceSpec extends AnyFlatSpec with MockitoSugar with Re
...
+      response should be(1)
+      consignmentIdCaptor.getValue should equal(expectedConsignmentId)
+      statusTypeCaptor.getValue should equal(nonUploadStatusType)
+      statusValueCaptor.getValue should equal(expectedStatusValue)
+      modifiedTimeCaptor.getValue should equal(modifiedTime)
+    }
+  }
+
+  forAll(statusTypes.filter(_ != "Upload")) { nonUploadStatusType =>
+    "updateConsignmentStatus" should s"throw an InputDataException when only a consignmentId and a non-'Upload' statusType of ${nonUploadStatusType} has been passed in" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala:472]: Maybe re-word test title slightly to: `... when a consignmentId and a non-'Upload' ...`

File: src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala [hunk 9]
```diff
@@ -350,13 +380,261 @@ class ConsignmentStatusServiceSpec extends AnyFlatSpec with MockitoSugar with Re
...
+        consignmentService.updateConsignmentStatus(updateConsignmentStatusInput).futureValue
+      }
+
+      thrownException.getMessage should equal(
+        s"Invalid ConsignmentStatus input: either '${nonUploadStatusType}' or ''"
+      )
+    }
+  }
+
+  "updateConsignmentStatus" should s"generate a 'Completed' statusValue when only a consignmentId and 'Upload' statusType have been passed in " +
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala:489]: Maybe re-word test title slightly to: `... when a consignmentId and 'Upload' ... and all consignment files have an ...`

File: src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala [hunk 10]
```diff
@@ -350,13 +380,261 @@ class ConsignmentStatusServiceSpec extends AnyFlatSpec with MockitoSugar with Re
...
+
+    val updateConsignmentStatusInput =
+      ConsignmentStatusInput(expectedConsignmentId, expectedStatusType, None)
+
+    consignmentService.updateConsignmentStatus(updateConsignmentStatusInput).futureValue
+
+    statusValueCaptor.getValue should equal(expectedStatusValue)
+  }
+
+  "updateConsignmentStatus" should s"generate a 'CompletedWithIssues' statusValue when only a consignmentId and 'Upload' statusType have been passed in " +
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala:530]: Maybe re-word test title slightly to: `... when a consignmentId and 'Upload' ... `

File: src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala [hunk 11]
```diff
@@ -350,13 +380,261 @@ class ConsignmentStatusServiceSpec extends AnyFlatSpec with MockitoSugar with Re
...
+    ).thenReturn(fileStatusMockRepoResponse)
+
+    val updateConsignmentStatusInput = ConsignmentStatusInput(expectedConsignmentId, expectedStatusType, None)
+
+    consignmentService.updateConsignmentStatus(updateConsignmentStatusInput).futureValue
+
+    statusValueCaptor.getValue should equal(expectedStatusValue)
+  }
+
+  "updateConsignmentStatus" should s"(given an Upload statusType) generate a statusValue based on the file status, even when a statusValue was passed in" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala:569]: Maybe re-word test title slightly to: `for Upload statusType ignore any statusValue pass in and generate the statusValue based on the file statuses of the consignment`

File: src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala [hunk 12]
```diff
@@ -350,13 +380,261 @@ class ConsignmentStatusServiceSpec extends AnyFlatSpec with MockitoSugar with Re
...
+      ConsignmentStatusInput(expectedConsignmentId, expectedStatusType, Some(passedInStatusValue))
+
+    consignmentService.updateConsignmentStatus(updateConsignmentStatusInput).futureValue
+
+    statusValueCaptor.getValue should equal(expectedStatusValue)
+  }
+
+  "updateConsignmentStatus" should s"throw an InputDataException if an Upload statusType was provided, a statusValue wasn't provided " +
+    "but none of the consignment's files had Upload statuses" in {
+    // This is just in case either, the client has forgotten to provide a statusValue or is trying to prematurely set
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala:612]: This comment isn't needed.

File: src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala [hunk 13]
```diff
@@ -350,13 +380,261 @@ class ConsignmentStatusServiceSpec extends AnyFlatSpec with MockitoSugar with Re
...
+
+    val updateConsignmentStatusInput =
+      ConsignmentStatusInput(expectedConsignmentId, expectedStatusType, Some(passedInStatusValue))
+
+    consignmentService.updateConsignmentStatus(updateConsignmentStatusInput).futureValue
+
+    statusValueCaptor.getValue should equal(expectedStatusValue)
+  }
+
+  "updateConsignmentStatus" should s"throw an InputDataException if an Upload statusType was provided, a statusValue wasn't provided " +
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusServiceSpec.scala:610]: Maybe reword the test title slightly to: `... for an Upload statusType with no statusValue where no consignment files have an Upload status`

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusService.scala [hunk 14]
```diff
@@ -1,22 +1,25 @@
...
 import java.sql.Timestamp
 import java.util.UUID
 import scala.concurrent.{ExecutionContext, Future}
 
 class ConsignmentStatusService(consignmentStatusRepository: ConsignmentStatusRepository,
+                               fileStatusRepository: FileStatusRepository,
                                uuidSource: UUIDSource,
                                timeSource: TimeSource)
                               (implicit val executionContext: ExecutionContext) {
+  private val statusesToUpdateBasedOnFile: Set[String] = Set("Upload")
```
[the remark this answers is not in the document]
Comment (techncl) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusService.scala:22]: Sure; looks cleaner. You mean `statusesToUpdateBasedOnFile.contains(input.statusType)` instead of `statusesToUpdateBasedOnFile.contains(statusValueToString)`, right?

**Review by TomJKing — COMMENTED** [review]
Just one very minor comment. Thanks

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentStatusRouteSpec.scala [hunk 15]
```diff
@@ -254,4 +254,93 @@ class ConsignmentStatusRouteSpec extends TestContainerUtils with Matchers with T
...
+      val expectedResponse = expectedUpdateConsignmentStatusMutationResponse("data_no_files_uploaded")
+      val response = runUpdateConsignmentStatusTestMutation("mutation_no_files_uploaded", token)
+
+      response.errors.head.message should equal(expectedResponse.errors.head.message)
+      response.errors.head.extensions should equal(expectedResponse.errors.head.extensions)
+  }
+
+  "updateConsignmentStatus" should "update the consignment status if statusType is 'Upload' and no statusValue was passed in, " +
+    "where all consignment files have 'Success' Upload statuses " +
+    "and files have 'Success' Upload statuses" in withContainers {
```
[the remark this answers is not in the document]
Comment (techncl) [reply src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentStatusRouteSpec.scala:297]: Oops 🤦 ; I'll remove it.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks
